### PR TITLE
Add benchmarking documentation

### PR DESCRIPTION
**Pull request by sebastienros**
Add benchmarking docs for Microsoft employees

### CODE REVIEW

File: docs/Benchmarks.md
Comment (captainsafia) [inline docs/Benchmarks.md:65]: Clarify that components here basically refers to the "project under test" and that you won't get benchmarks correctly if your code modification isn't in the component.

Comment (sebastienros) [reply docs/Benchmarks.md:65]: Added section explaining each part

File: docs/Benchmarks.md
Comment (halter73) [inline docs/Benchmarks.md:67]:
```suggestion
- `kestrel` (Microsoft.AspNetCore.Server.Kestrel `src\Servers\Kestrel\build.cmd`)
- `mvc` (Microsoft.AspNetCore.Mvc `src\Mvc\build.cmd`)
- `routing` (Microsoft.AspNetCore.Routing `src\Http\build.cmd`)

See [prbenchmarks.aspnetcore.config.yml](https://github.com/aspnet/Benchmarks/blob/2bddb9b43ffbc8cfdb4a2c3777d4e5213bc42332/build/prbenchmarks.aspnetcore.config.yml) for more details.
```

File: docs/Benchmarks.md
Comment (halter73) [inline docs/Benchmarks.md:76]: I think we should be even more clear that changes outside of the "component" projects will not be benchmarked like @captainsafia suggested. I think that's the biggest footgun because it's not obvious when you get it wrong.

```suggestion
- *Component* is the ASP.NET project under test that will be built and contains your changes.

> [!IMPORTANT]  
> Any changes outside of the "component" projects will not be benchmarked.
```

**Review by halter73 — APPROVED** [review]

File: docs/Benchmarks.md
Comment (adityamandaleeka) [inline docs/Benchmarks.md:3]:
```suggestion
This document explains how to benchmark ASP.NET Core changes. Currently, only members of the .NET team at Microsoft have the permissions necessary to trigger performance runs on the official infrastructure.

```

File: docs/Benchmarks.md
Comment (adityamandaleeka) [inline docs/Benchmarks.md:90]: Is there an issue for this that can be linked here?

File: docs/Benchmarks.md
Comment (adityamandaleeka) [inline docs/Benchmarks.md:3]: Should this say that the document explains how to benchmark changes _in the aspnetcore perf lab_?

File: docs/Benchmarks.md
Comment (sebastienros) [reply docs/Benchmarks.md:90]: Currently adding more log in crank, you can approve the PR if you want to help. Right now there is no exception but nothing is in the comments either.

Comment (adityamandaleeka) [reply docs/Benchmarks.md:90]: I helped! 😆

File: docs/Benchmarks.md
Comment (danmoseley) [reply docs/Benchmarks.md:3]: > Currently, only members of the .NET team at Microsoft

We usually just say "the .NET team".

Aside -- at some future point I'd be curious to understand how close we are to being able to allow other contributors to run these jobs. ie., whether there's anything needs to be more robust to make it merely a resource/cost issue that we could tackle by maintaining a trusted group. (We can discuss offline sometime)

File: docs/Benchmarks.md
Comment (danmoseley) [inline docs/Benchmarks.md:4]: Consider a section here describing what PR's we imagine should use such runs? Essentially -- what kinds of changes tend to lead to perf regressions (that would be caught) and what would not benefit. This may also depend on how heavily the machines are subscribed -- that could feed into guidance here.

(If/when we set up a bot to post an encouragement/instructions into PR's that would most benefit from a perf run, perhaps by looking at files changed, that will obviously also help. Note that the labeler bot, eg., already looks at files changed in a PR.)

File: docs/Benchmarks.md
Comment (danmoseley) [inline docs/Benchmarks.md:12]: Maybe briefly explaining what Crank is -- what it can do, what it supports (eg., architectures), how long jobs typically take, ..

Personally I'm also curious where the name came from...

File: docs/Benchmarks.md
Comment (danmoseley) [inline docs/Benchmarks.md:43]: maybe link to a past PR that requested some runs this way, as an example?

File: docs/Benchmarks.md
Comment (danmoseley) [inline docs/Benchmarks.md:90]: how do I know whether the results represent a regression? does it automatically do a baseline on HEAD at the same time? or i have to go compare/interpret the dashboard (that seems error prone given the complexity of the dashboard)

does it help me understand whether a diff might  be noise?

File: docs/Benchmarks.md
Comment (danmoseley) [inline docs/Benchmarks.md:3]: I don't see anything about microbenchmarks 
1. the ones checked into dotnet/aspnetcore that are never run automatically
2. the ones in dotnet/performance -- do any of those run through code that exists in dotnet/aspnetcore

If crank supports any of these, and I hear it now supports dotnet/performance, we should link to instructions.
for the ones it doesn't support, we should at least point out they exist, and ideally link to instructions 

this is where the commentary on how to choose relevant benchmarks might come in handy. let's try to couch these instructions in terms that a new hire can use, who doesn't know about any of these things.

File: docs/Benchmarks.md
Comment (danmoseley) [inline docs/Benchmarks.md:5]: maybe here have a note that you're going to describe two alternatives -- local and PR -- below. they should choose which.

File: docs/Benchmarks.md
Comment (danmoseley) [inline docs/Benchmarks.md:38]: Is there anything I don't get from a PR run that I do get from a local run?

File: docs/README.md
Comment (danmoseley) [inline docs/README.md:32]:
```suggestion
| [Benchmarking](Benchmarks.md) | Instructions on how to benchmark PRs and local changes | .NET team |
```

File: docs/Benchmarks.md
Comment (danmoseley) [inline docs/Benchmarks.md:90]: > After up to 10 minutes a link should be displayed

what should I do if it isn't, or I hit some other trouble? look at dotnet/crank docs? use some teams chat? (we're hopefully getting away from "ping Seb first thing" :))

File: docs/Benchmarks.md
Comment (danmoseley) [reply docs/Benchmarks.md:3]: particularly since this doc is titled "how to benchmark changes" not "how to do system level benchmark of changes"

Comment (sebastienros) [reply docs/Benchmarks.md:3]: This is a hard requirement since jobs are triggered on dnceng/internal. Allowing any external user to run anything in this ring would compromise secrets. So we only accept users that have the merge permission on this repository as this would have the same effect. So I would believe that this will be possible when we allow external contributors to merge PRs to main or release/x.

Comment (sebastienros) [reply docs/Benchmarks.md:3]: The micro benchmarks in this repository were made mostly for investigations and not regression detections.

File: docs/Benchmarks.md
Comment (sebastienros) [reply docs/Benchmarks.md:4]: I'd rather keep this document as succint as possible for devs who want to run a benchmark but don't remember how to. My personal opinion for which PRs should use such run is only "for external contributor PRs", and have .NET devs use locally built branches instead. I mentioned that in the dedicated paragraph.

File: docs/Benchmarks.md
Comment (danmoseley) [reply docs/Benchmarks.md:3]: Gotcha

File: docs/Benchmarks.md
Comment (danmoseley) [reply docs/Benchmarks.md:5]: OK

File: docs/Benchmarks.md
Comment (sebastienros) [reply docs/Benchmarks.md:38]: If you are actively working on this PR you will need to push your changes and then wait more that 10 minutes for every change. Locally started benchmarks are much quicker to get results from, and don't need to push your code on GH. You also can't get profiles.

**Pull request merged**
